pkg/util/cost_attribution.go

```diff
@@ -0,0 +1,185 @@
+	if groupTs := ag.timestampsPerUser[userID][attribution]; groupTs != nil {
+		ag.mu.RUnlock()
+		groupTs.Store(ts)
+		return
+	}
+	ag.mu.RUnlock()
+
+	ag.mu.Lock()
+	defer ag.mu.Unlock()
+
+	if ag.timestampsPerUser[userID] == nil {
+		ag.timestampsPerUser[userID] = map[string]*atomic.Int64{attribution: atomic.NewInt64(ts)}
+		return
+	}
+
+	if groupTs := ag.timestampsPerUser[userID][attribution]; groupTs != nil {
+		groupTs.Store(ts)
+		return
+	}
+
+	ag.timestampsPerUser[userID][attribution] = atomic.NewInt64(ts)
+}
+
+func (ag *CostAttribution) purgeInactiveAttributionsForUser(userID string, deadline int64) []string {
+	ag.mu.RLock()
+	var inactiveAttributions []string
+	attributionTimestamps := ag.timestampsPerUser[userID]
+
+	for attr, ts := range attributionTimestamps {
+		if ts.Load() <= deadline {
+			inactiveAttributions = append(inactiveAttributions, attr)
+		}
+	}
+	ag.mu.RUnlock()
+
+	if len(inactiveAttributions) == 0 {
+		return nil
+	}
+
+	// Cleanup inactive groups
+	ag.mu.Lock()
+	defer ag.mu.Unlock()
+
+	for i := 0; i < len(inactiveAttributions); {
+		inactiveAttribution := inactiveAttributions[i]
+		groupTs := ag.timestampsPerUser[userID][inactiveAttribution]
+		if groupTs != nil && groupTs.Load() <= deadline {
+			delete(ag.timestampsPerUser[userID], inactiveAttribution)
+			i++
+		} else {
+			inactiveAttributions[i] = inactiveAttributions[len(inactiveAttributions)-1]
+			inactiveAttributions = inactiveAttributions[:len(inactiveAttributions)-1]
+		}
+	}
+
+	return inactiveAttributions
+}
+
+func (ca *CostAttribution) purgeInactiveAttributions(inactiveTimeout time.Duration, cleanupFuncs ...func(string, string)) {
+	ca.mu.RLock()
+	userIDs := make([]string, 0, len(ca.timestampsPerUser))
+	for userID := range ca.timestampsPerUser {
+		userIDs = append(userIDs, userID)
+	}
+	ca.mu.RUnlock()
+
+	currentTime := time.Now()
+	for _, userID := range userIDs {
+		inactiveAttributions := ca.purgeInactiveAttributionsForUser(userID, currentTime.Add(-inactiveTimeout).UnixNano())
+		for _, attribution := range inactiveAttributions {
+			for _, cleanupFn := range cleanupFuncs {
+				cleanupFn(userID, attribution)
+			}
+		}
+	}
+}
+
+func (ca *CostAttribution) attributionLimitExceeded(userID, attribution string, now time.Time) bool {
+	// if we are still at the cooldown period, we will consider the limit reached
+	ca.mu.RLock()
+	defer ca.mu.RUnlock()
+
+	if v, exists := ca.coolDownDeadline[userID]; exists && v.Load() > now.UnixNano() {
+		return true
+	}
+
+	// if the user attribution is already exist and we are not in the cooldown period, we don't need to check the limit
+	_, exists := ca.timestampsPerUser[userID][attribution]
+	if exists {
+		return false
+	}
+
+	// if the user has reached the limit, we will set the cooldown period which is 20 minutes
+	maxReached := len(ca.timestampsPerUser[userID]) >= ca.maxAttributionPerUser
+	if maxReached {
+		ca.coolDownDeadline[userID].Store(time.Now().Add(20 * time.Minute).UnixNano())
```

Comment:
Cool down interval should be configurable.
Shouldn't it be the same as inactiveTimeout?

In that case, I think we can just delete all user labels here, because we know they all are going to expire during the cool down period.

Reply:
added cooldown config